### PR TITLE
Include tlstunnel filesystem

### PR DESCRIPTION
**Pull request by reynir**
The interface differs quite a bit

### CODE REVIEW

**Comment by reynir** [comment]
I spent some time preserving the filesystem.ml history from tlstunnel - hence quite a few commits by hannes.

**Comment by hannesm** [comment]
Great to see this. I'm curious what should the name be? Is `lossless` good? Maybe redundant? or fault-tolerant?

And should we try to work on a migration from "lossy" to "lossless"? (And/or the other way around?)

I guess in the mli we should describe the different approaches.

**Comment by reynir** [comment]
Maybe better names are indeed `cache` and `fault_tolerant`. I will rename.

What do you think of the `read` function? I struggle a bit what to do with the in-memory superblock and the read superblock - should we compare them?! Bar race conditions they should be exactly the same. I was considering not implementing `read` because of this.

**Comment by reynir** [comment]
Also another thought! Should we split these into two sub libraries? So a user of the lossy/cache version doesn't link digestif, and vice versa for checkseum? I think that would be nice.

**Comment by hannesm** [comment]
I'm in favour of spliiting into sub-libraires... or maybe even separate opam packages, I'm not sure...

If we consider writing "convert lossy to lossless" or "convert lossless to lossy", a single opam package makes perfect sense.

`read` let me try to get my head around that

File: lib/oneFFS.ml [hunk 1]
```diff
@@ -163,3 +163,50 @@ module Make(B : Mirage_block.S) = struct
...
+    mutable superblock : FS.superblock;
+  }
+
+  let connect b =
+    let* info = Block.get_info b in
+    if info.sector_size < FS.superblock_size then
+      failwith "sector size too small for this filesystem";
+    let* r = FS.read_data b in
+    match r with
+    | Error `Bad_checksum ->
```
Comment (hannesm) [inline lib/oneFFS.ml:183]: not 100% sure this behaviour is intended... so we read something bad from the disk... but does that mean we want to initialize and return "we have read nothing"? if a user has a bad disk with broken blocks, shouldn't that at least be mentioned in a log message that the checksum verification failed (so the user can try to find a backup)?

File: lib/filesystem.ml [hunk 2]
```diff
@@ -0,0 +1,240 @@
...
+  let lwt_err_to_msg ~pp_error f =
+    let open Lwt.Infix in
+    f >|= Result.map_error (fun e -> `Msg (Fmt.to_to_string pp_error e))
+
+  let read_data block =
+    let open Lwt.Infix in
+    Block.get_info block >>= fun info ->
+    let open Lwt_result.Infix in
+    let ss = info.Mirage_block.sector_size in
+    assert (ss >= superblock_size);
```
Comment (hannesm) [inline lib/filesystem.ml:109]: TODO: we should get rid of assertions in the code. log or return error...

Comment (reynir) [reply lib/filesystem.ml:109]: I saw this and added a check in `connect`. We can safely remove this `assert`.
```suggestion
```

File: lib/oneFFS.ml [hunk 3]
```diff
@@ -163,3 +163,50 @@ module Make(B : Mirage_block.S) = struct
...
+
+  let write t data =
+    let+ r = FS.write_data t.b t.superblock data in
+    Result.map (fun superblock -> t.superblock <- superblock) r
+
+  let read t =
+    let+ r = FS.read_data t.b in
+    match r with
+    | Ok (superblock, data) ->
+      (* XXX: what to do with the read superblock?! *)
```
Comment (hannesm) [inline lib/oneFFS.ml:206]: so in the tlstunnel unikernel, we do:

```OCaml
FS.read_data block >>= function
...
    | Ok (superblock, data) ->
      let config = { superblock ; sni = Domain_name.Host_map.empty } in
...

   let write_configuration block config =
    let open Lwt_result.Infix in
    let data = Configuration.encode_data config.sni in
    FS.write_data block config.superblock data >|= fun superblock ->
    config.superblock <- superblock
```

So, it looks for writing we need the superblock. But we already have that from `connect`... so a read is pretty superfluous?!? i.e. when would you want to read again? certainly you read at startup, and then when you change something you write, but you should never need to read again.

But for the XXX: we can just ignore the superblock :)

File: lib/oneFFS.ml (hunk 1 above)
Comment (reynir) [reply lib/oneFFS.ml:183]: Hm, yes. I will add a log message.

FWIW `FS.read_data` will try to fix a broken superblock if the other one passes checksum. I took this code from tlstunnel. It seems `` `Bad_checksum `` is returned also on an uninitialized disk - so the intent was maybe to initialize the disk if it wasn't? I can add a separate `format` function to do that.

Comment (hannesm) [reply lib/oneFFS.ml:183]: That's right, it is as well used for formatting... I think a log message is fine -- the API client receives `None` as value, so if they expected data they should do something about it ;)

File: lib/oneFFS.ml (hunk 3 above)
Comment (reynir) [reply lib/oneFFS.ml:206]: Yes, exactly! I would only read again if for some reason I decide not to keep the data in memory. But the format doesn't lend well to partial reads, so you wouldn't store more data than you can keep in memory.

**Comment by reynir** [comment]
I think this is good to merge now. There may be more things that would be nice to do before publishing, such as moving to the robur-coop organization (#4). I'm also thinking if it's desirable to have a on-disk representation for initialized-but-no-data. Currently, `init` initializes with data being the empty string. I'll open an issue for that...

**Comment by hannesm** [comment]
> on-disk representation for initialized-but-no-data

I wouldn't bother with that. If we want it, we can just write the empty string :)

**Pull request merged**